rclcpp/include/rclcpp/parameter.hpp

```diff
@@ -28,6 +28,25 @@
 namespace rclcpp
 {
 
+class Parameter;
+
+namespace node_interfaces
+{
+struct ParameterInfo;
+}  // namespace node_interfaces
+
+namespace detail
+{
+
+// This helper function is required because you cannot do specialization on a
+// class method, so instead we specialize this template function and call it
+// from the unspecialized, but dependent, class method.
+template<typename T>
+auto
```

Comment:
I don't know why auto return value is inherently bad, but I don't know of another way to avoid it. @sloretz wrote this part I believe, but I think it's correct.

Reply:
I wouldn't say "inherently bad", but it does make it more difficult for the reader later to know which type is returned (the most common place I run into this is when trying to make a `std::vector` of what's returned here, and I need to go grubbing around in the source to find what is actually returned).  So generally I prefer to only use auto when the type is glaringly obvious.

Reply:
Maybe the return type could be `T`? I'm not sure, I think the auto here is to avoid redundancy with the function it calls (it's more of a forwarding of template arguments). Like if the called function (in the implementation of this function) returns int64_t when T = int, then you'd want to avoid the logic needed to do that here as well, therefore auto.

I'd prefer not to muck with this anymore, as I had to do some stuff with `decltype(auto)` and things like that to make it work.

It would all be a lot simpler with constexpr-if, though I'd still need auto return type in that case.